docs/design/uptime-reputation-calculation.md

```diff
@@ -0,0 +1,122 @@
+
+When the _SN_ is disqualified by any disqualification system and the _SN_ isn't back online, the `disqualified` column of the row will be updated for not receiving more uptime checks because we don't do business with disqualified _SNs_.
+
+### Algorithm
+
+The algorithm is composed of 2 parts:
+
+1. Select which _SNs_ will be checked for uptime, henceforth uptime check _SN_ selection.
+1. Recheck the uptime of the _SNs_ which has failed the initial uptime check (part 1) during an interval period of time, henceforth uptime recheck loop.
+
+#### Uptime Check _SN_ selection
+
+This design doc doesn't change the _SN_ selection process for uptime checks therefore the current implemented mechanism with some minor modifications to classify those nodes which fail the uptime check in order of being recheck by the second part of the algorithm.
+
+Once an _SN_ is selected for an uptime check, the current implemented uptime check mechanism must be modified to achieve the following algorithm:
+
+1. If there is a row in the `failed_uptime_checks` table for the selected _SN_ with `back_online` set to `NULL` do nothing, otherwise follow with the next step.
+1. Check the _SN_ as it's currently done; if it succeeds, then end, otherwise follow with the next step.
+1. Insert a new row in the `failed_uptime_checks` using its node ID and setting the current timestamp to `when`.
+
+
+#### Uptime Recheck Loop
+
+This process should run independently from any current satellite process and it should run in a configurable time interval.
+
+The algorithm for each time interval iteration is the following:
+
+1. Select all the rows from `failed_uptime_checks` table which has `back_online` and `disqualification` columns set to `NULL` sorted by `when` in ascending order.
+1. For each selected _SN_, retrieve the _SN_ address<sup>1</sup> and performs the uptime check.
+    * If it fails
+        1. Calculate its total offline time of the month, accumulating the offline time interval of each row corresponding to this _SN_ with `back_online` not being `NULL` and the current offline interval time calculated with the current timestamp and the `when` value of this row. If the total offline time doesn't exceed the established network limits end, otherwise follow with the next step.
+        2. Update the `disqualified` column of the row with the current timestamp and disqualifies the _SN_<sup>2</sup>
+    * If it succeeds
+        1. Update the `back_online` column of the row with the current timestamp.
+
+
+## Open issues
+
+Currently we have the following issues and concerns which can affects somehow the implementation of this system:
+
+1. Currently the uptime checks are combined with the audit check for calculating the total _SN_ reputation; after this implementation, the uptime check _"score"_ cannot be combined anymore as before.
```

Comment:
I would say combined. Uptime checks and audit checks are working independent. Storage nodes can get disqualified for beeing offline too long or for failing too many audits. They can't get disqualified for a combination of both. One of them has to hit the magic line.

Reply:
For me that's OK but we'll have to consult data science team about it because the initial implementation and the one that we have now (although it's disabled) is mentioning to combine them.

Reply:
we combine them when creating an overall reputation score by which we prioritize node selection using "the power of two choice"

Reply:
That's a good point; I didn't remember so.

Then it looks that we have to look for a way to get a value that we can combine them.

Reply:
This has been discussed with @benjaminsirb and the document has been updated.